checker/jdk/determinism/src/java/util/TreeMap.java

```diff
@@ -0,0 +1,3021 @@
+{
+    /**
+     * The comparator used to maintain order in this tree map, or
+     * null if it uses the natural ordering of its keys.
+     *
+     * @serial
+     */
+    private final Comparator<? super K> comparator;
+
+    private transient Entry<K,V> root;
+
+    /**
+     * The number of entries in the tree
+     */
+    private transient int size = 0;
+
+    /**
+     * The number of structural modifications to the tree.
+     */
+    private transient int modCount = 0;
+
+    /**
+     * Constructs a new, empty tree map, using the natural ordering of its
+     * keys.  All keys inserted into the map must implement the {@link
+     * Comparable} interface.  Furthermore, all such keys must be
+     * <em>mutually comparable</em>: {@code k1.compareTo(k2)} must not throw
+     * a {@code ClassCastException} for any keys {@code k1} and
+     * {@code k2} in the map.  If the user attempts to put a key into the
+     * map that violates this constraint (for example, the user attempts to
+     * put a string key into a map whose keys are integers), the
+     * {@code put(Object key, Object value)} call will throw a
+     * {@code ClassCastException}.
+     */
+    public @Det TreeMap() {
+        comparator = null;
+    }
+
+    /**
+     * Constructs a new, empty tree map, ordered according to the given
+     * comparator.  All keys inserted into the map must be <em>mutually
+     * comparable</em> by the given comparator: {@code comparator.compare(k1,
+     * k2)} must not throw a {@code ClassCastException} for any keys
+     * {@code k1} and {@code k2} in the map.  If the user attempts to put
+     * a key into the map that violates this constraint, the {@code put(Object
+     * key, Object value)} call will throw a
+     * {@code ClassCastException}.
+     *
+     * @param comparator the comparator that will be used to order this map.
+     *        If {@code null}, the {@linkplain Comparable natural
+     *        ordering} of the keys will be used.
+     */
+    public @PolyDet TreeMap(@PolyDet Comparator<? super K> comparator) {
+        this.comparator = comparator;
+    }
+
+    /**
+     * Constructs a new tree map containing the same mappings as the given
+     * map, ordered according to the <em>natural ordering</em> of its keys.
+     * All keys inserted into the new map must implement the {@link
+     * Comparable} interface.  Furthermore, all such keys must be
+     * <em>mutually comparable</em>: {@code k1.compareTo(k2)} must not throw
+     * a {@code ClassCastException} for any keys {@code k1} and
+     * {@code k2} in the map.  This method runs in n*log(n) time.
+     *
+     * @param  m the map whose mappings are to be placed in this map
+     * @throws ClassCastException if the keys in m are not {@link Comparable},
+     *         or are not mutually comparable
+     * @throws NullPointerException if the specified map is null
+     */
+    public @PolyDet TreeMap(@PolyDet Map<? extends K, ? extends V> m) {
```

Comment:
In this case, we now have code that replaces `@OrderNonDet` with `@Det`:

https://github.com/t-rasmud/checker-framework/blob/ef64b4fc12c2ec45df2a2c635a7d3c3d62a5fc20/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java#L319

Reply:
I understand having the checks there to prevent explicit construction of a `@PolyDet` TreeSet, but the code to replace `@OrderNonDet` seems redundant when we could instead just make the return type of this constructor `@PolyDet("up")`. Either way, it would be good for the annotations to agree with the hardcoded behavior as much as possible.

Reply:
We are actually replacing `@OrderNonDet` with `@Det` here. So `@PolyDet("down")` seems to be the more appropriate choice here. Additionally, the modifiers "up" and "down" on `@PolyDet` have no effect if the underlying type is a collection, a map or an array.